language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/json/JsonLfReader.java

```diff
@@ -0,0 +1,443 @@
+      String value = text().read();
+      try {
+        return Enum.valueOf(enumClass, value);
+      } catch (IllegalArgumentException e) {
+        parseExpected(String.format("constant of %s", enumClass.getName()));
+      }
+      return null;
+    };
+  }
+
+  // Provides a generic way to read a variant type, by specifying each tag.
+  public <T> FromJson<T> variant(List<String> tagNames, TagReader<T> readTag) {
+    return () -> {
+      readStartObject();
+      if (!readFieldName().equals("tag")) parseExpected("tag field");
+      String tagName = text().read();
+      if (!readFieldName().equals("value")) parseExpected("value field");
+      T result = readTag.get(tagName);
+      readEndObject();
+      if (result == null) parseExpected(String.format("tag of %s", String.join(" or ", tagNames)));
+      return result;
+    };
+  }
+
+  public interface TagReader<T> {
+    T get(String tagName) throws FromJson.Error;
+  }
+
+  // Provides a generic way to read a record type, with a constructor arg for each field.
+  // This is a little fragile, so is better used by code-gen. Specifically:
+  // - The constructor must cast the elements and pass them to the T's constructor appropriately.
+  // - The elements of fieldNames should all evaluate to non non-null when applied to fieldsByName.
+  // - The argIndex field values should correspond to the args passed to the constructor.
+  //
+  // e.g.
+  //     r.record(
+  //        args -> new Foo((Long) args[0], (Boolean) args[1]),
+  //        asList("i", "b"),
+  //        fieldName -> {
+  //          switch (fieldName) {
+  //            case "i":
+  //              return JsonLfReader.Field.required(0, r.int64());
+  //            case "b":
+  //              return JsonLfReader.Field.optional(1, r.bool(), false);
+  //            default:
+  //              return null;
+  //          }
+  //        }
+  //     )
+  public <T> FromJson<T> record(
+      Function<Object[], T> constr,
+      List<String> fieldNames,
+      Function<String, Field<? extends Object>> fieldsByName) {
+    return () -> {
+      List<String> missingFields = new ArrayList<>();
+      List<String> unknownFields = new ArrayList<>();
+
+      Object[] args = new Object[fieldNames.size()];
+      if (isStartObject()) {
+        readStartObject();
+        while (notEndObject()) {
+          String fieldName = readFieldName();
+          var field = fieldsByName.apply(fieldName);
+          if (field == null) unknownFields.add(fieldName);
+          else args[field.argIndex] = field.fromJson.read();
+        }
+        readEndObject();
+      } else if (isStartArray()) {
+        readStartArray();
+        for (String fieldName : fieldNames) {
+          var field = fieldsByName.apply(fieldName);
+          args[field.argIndex] = field.fromJson.read();
+        }
+        readEndArray();
+      } else {
+        parseExpected("object or array");
+      }
+
+      // Handle missing and unknown fields.
```

Comment:
generally JSON decoders tend to be lax wrt to unknown fields. This also fits in nicely with the new upgrading proposal. I would suggest to be lax here from the start.

Reply:
seems fine to fail for now and revisit this once we get to the new upgrading proposal.